### PR TITLE
ARROW-11044: [C++] Add "replace" kernel

### PR DESCRIPTION
**Pull request by bu2**
Purpose a "replace" compute kernel which could fulfil ~~ARROW-10641 - [C++] A "replace" or "map" kernel to replace values in array based on mapping (@jorisvandenbossche)~~ [ARROW-9430](https://issues.apache.org/jira/browse/ARROW-9430) - [C++/Python] Kernel for SetItem(BooleanArray, values) (@xhochy). The implementation started on the basis of "fill_null" kernel. But it takes an additional BooleanArray parameter which is used as a mask to trigger value replacement.

With following "null handling" logic:

```
Values		Mask		Replacement	Output
1 v		1 false		1 r		1 v
1 v		1 true		1 r		1 r
1 v		0 null		1 r		1 v
1 v		1 false		0 null		1 v
1 v		1 true		0 null		1 v
1 v		0 null		0 null		1 v
0 null		1 false		1 r		0 null
0 null		1 true		1 r		1 r
0 null		0 null		1 r		0 null
0 null		1 false		0 null		0 null
0 null		1 true		0 null		0 null
0 null		0 null		0 null		0 null
```
(each column indicates the validity bit and the corresponding value. "v" for current value, "r" for replacement value).

### CODE REVIEW

**Comment by bu2** [comment]
@jorisvandenbossche: Would you have you look at this in-progress PR and share your thought ? Replace kernel implemented for most (if not all) types but missing null handling.

**Comment by bu2** [comment]
**Note on ArrayData::GetValues<Type>(i)** in the context of commit https://github.com/apache/arrow/pull/9024/commits/ab2d3d2e0c4b91e54b808a886364d4c0cfbb19ae:

* ArrayData::GetValues<T>(i) use "byte offset" even for buffers[2] which stores variable width values...
=> Should it rather use buffers[1] value offset ?
=> Or force absolute_offset to 0 for buffers[2] ?

On my case, I fixed the bug in my code by forcing absolute_offset to 0.

File: cpp/src/arrow/compute/kernels/scalar_replace.cc [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+// under the License.
+
+#include "arrow/compute/kernels/common.h"
+#include "arrow/util/bit_block_counter.h"
+#include "arrow/util/bit_util.h"
+#include "arrow/util/bitmap_ops.h"
+
+namespace arrow {
+
+using internal::BitBlockCount;
```
Comment (emkornfield) [inline cpp/src/arrow/compute/kernels/scalar_replace.cc:25]: use full imports and put them inside the anonymous namespace.

**Comment by pitrou** [comment]
I wouldn't expect a "replace" operation to do this. Instead, this looks more like a "select" operation. @wesm What do you think?

**Comment by jorisvandenbossche** [comment]
This might actually rather be the "setitem" kernel as proposed in https://issues.apache.org/jira/browse/ARROW-9430 ? (which is something we want as well)

**Comment by pitrou** [comment]
"setitem" is confusing to me (I would expect something where you pass indices and values to set at those indices). "select" is used in [Numpy](https://numpy.org/doc/stable/reference/generated/numpy.select.html). Another possible is "cond".

**Comment by jorisvandenbossche** [comment]
> "setitem" is confusing to me (I would expect something where you pass indices and values to set at those indices).

That's https://issues.apache.org/jira/browse/ARROW-9431. 
There could be a variant of "setitem" that takes integer indices, and one that takes a boolean mask?

But it's true that for a "setitem" kernel, you would expect the "values to set" (`Replacement` in the example in the top post) to be of the length of the number of values that get set, not of the same length as the original array.

So what this PR proposes, is actually more a "where" kernel (using numpy's terminology), for which we have https://issues.apache.org/jira/browse/ARROW-10640. 
There is some discussion on that PR that this can be generalized to something like numpy's `np.choose`. And the `np.select`  mentioned by @pitrou is quite related to that as well (the exact semantics are a bit different between both: `select` uses boolean conditions to choose between the arrays, `choose` a single array of indices into the choice arrays)

**Comment by pitrou** [comment]
I'd say that personally, I find "select" more intuitively guessable than "where" (which is a weird name for a function). But I could also live with it if it's Numpy's choice. @michalursa

**Comment by jorisvandenbossche** [comment]
Agreed that "where" is an unintuitive name. For me it's fine to directly use the more general name from numpy, which would actually be both "choose" or "select" (I think the single boolean mask can be seen as a special case for both generalized functions: assuming that the single boolean mask from "where" is interpreted as 0/1 integer indices) for "choose", or is interpreted as a boolean mask + its inverse mask for "select"). 

Another option mentioned in the JIRA is "if_else" (term used in the R world) or "case_when" (inspired on SQL).

**Comment by pitrou** [comment]
"if_else", "choose", "select" are all fine with me.

**Comment by bu2** [comment]
Wow, I was not expecting such brainstorming on the name of this kernel. My initial intent was to mimic pandas.Series.replace(to_replace, value), where in C++ to_replace could be an BooleanArray mask (same length as input array) which trigger value replacement => So this kernel could be used after any combination of Compare and Boolean operations to implement specific replacement logic.

This could be extended to another implementation where to_replace could be an IntegerArray (length <= as input array) of indexes to replace.

Here is how I use "replace" in combination with "is_nan" ([ARROW-11043](https://github.com/apache/arrow/pull/9023)) to implement fillna():

```
template <typename value_type>
std::shared_ptr<DataFrame> DataFrame::fillna(value_type value) {
    auto outdf = std::make_shared<DataFrame>();
    ...
    ...
    for (int i = 0 ; i < this->table_->num_columns() ; ++i) {
        auto field = this->table_->schema()->field(i);
        auto chunked_array = this->table_->column(i);
        auto value_datum = arrow::compute::Cast(arrow::Datum(value), chunked_array->type()).ValueOrDie();
        auto nulls = arrow::compute::IsNull(chunked_array).ValueOrDie();
        auto nans = arrow::compute::IsNan(chunked_array).ValueOrDie().chunked_array();
        auto to_replace = arrow::compute::Or(nulls, nans).ValueOrDie();
        auto clean_chunked_array = arrow::compute::Replace(chunked_array, nans, value_datum).ValueOrDie().chunked_array();
        outdf->table_ = outdf->table_->AddColumn(i, field, clean_chunked_array).ValueOrDie();
    }

    return outdf;
}
```

File: cpp/src/arrow/compute/kernels/scalar_replace.cc (hunk 1 above)
Comment (bu2) [reply cpp/src/arrow/compute/kernels/scalar_replace.cc:25]: Sorry @emkornfield, not sure to get your point... Do you mean:

```
namespace arrow {
namespace compute {
namespace internal {

namespace {

using ::arrow::internal::BitBlockCount;
using ::arrow::internal::BitBlockCounter;

...
...
}

}
}
}
```
:question:

**Comment by pitrou** [comment]
> Wow, I was not expecting such brainstorming on the name of this kernel.

Sorry for the distraction :-) That said, for clarity it seems that "if_else" would be both explicit and non-misleading.

The signature could be `if_else(cond, when_true, when_false)` with the semantics `cond ? when_true : when_false`. Nulls would probably just pass through.

Do you want to update this PR?

**Comment by jorisvandenbossche** [comment]
@bu2 would you have time to update this based on the above comments? (https://github.com/apache/arrow/pull/9024#issuecomment-775323925)

File: cpp/src/arrow/compute/kernels/scalar_replace.cc [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+          for (int64_t i = 0; i < block.length; ++i) {
+            out_values[i] =
+                BitUtil::GetBit(to_replace, offset + i) ? value : in_values[i];
+          }
+        }
+        offset += block.length;
+        out_values += block.length;
+        in_values += block.length;
+      }
+      output->buffers[1] = out_buf;
```
Comment (nirandaperera) [inline cpp/src/arrow/compute/kernels/scalar_replace.cc:104]: minor comment, you could use `std::move` here. (applies to several places, where buffer shared_ptr's are assigned)

**Comment by pitrou** [comment]
Closing as superceded by #10410.

**Pull request closed without merging**